src/legacy/ui/public/vis/editors/default/agg_param.js

```diff
@@ -1,136 +1,140 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { isFunction } from 'lodash';
+import { wrapInI18nContext } from 'ui/i18n';
+import { uiModules } from '../../../modules';
+import { AggParamReactWrapper } from './agg_param_react_wrapper';
+
+uiModules
+  .get('app/visualize')
+  .directive('visAggParamReactWrapper', reactDirective => reactDirective(wrapInI18nContext(AggParamReactWrapper), [
+    ['agg', { watchDepth: 'collection' }],
+    ['aggParam', { watchDepth: 'reference' }],
+    ['indexedFields', { watchDepth: 'collection' }],
+    ['paramEditor', { wrapApply: false }],
+    ['onChange', { watchDepth: 'reference' }],
+    ['setTouched', { watchDepth: 'reference' }],
+    ['setValidity', { watchDepth: 'reference' }],
+    'field',
+    'isInvalid',
+    'value',
+    'otherBucket',
+    'missingBucket',
+  ]))
+  .directive('visAggParamEditor', function (config) {
+    return {
+      restrict: 'E',
+      // We can't use scope binding here yet, since quiet a lot of child directives arbitrary access
+      // parent scope values right now. So we cannot easy change this, until we remove the whole directive.
+      scope: true,
+      require: '?^ngModel',
+      template: function ($el, attrs) {
+        if (attrs.editorComponent) {
+          // Why do we need the `ng-if` here?
+          // Short answer: Preventing black magic
+          // Longer answer: The way this component is mounted in agg_params.js (by manually compiling)
+          //   and adding to some array, once you switch an aggregation type, this component will once
+          //   render once with a "broken state" (something like new aggParam, but still old template),
+          //   before agg_params.js actually removes it from the DOM and create a correct version with
+          //   the correct template. That ng-if check prevents us from crashing during that broken render.
+          return `<vis-agg-param-react-wrapper
+            ng-if="editorComponent"
+            param-editor="editorComponent"
+            agg="agg"
+            agg-param="aggParam"
+            field="agg.params.field"
+            indexed-fields="indexedFields"
+            is-invalid="isInvalid"
+            value="paramValue"
+            on-change="onChange"
+            set-touched="setTouched"
+            set-validity="setValidity"
+            other-bucket="agg.params.otherBucket"

```

Comment:
I don't like we're now manually specifying all the parameters we want to listen for. That sounds like a really dirty workaround. I am wondering, if we can't simply switch the `agg` property above from `watchDepth: 'collection'` to `watchDepth: 'value'` (or just skip it then, since that's the default). I remember that there has been some kind of infinite loop by that maybe, but we could check if we can at least just add one new `params="agg.params"` with a watch depth of value, so this will automatically rerender on all parameter value changes.

Reply:
This comment is addressed here #35153